File: phases/04-algorithms-and-complexity/21-hashing-in-algorithms-rabin-karp-rolling-hashes/code/main.py
"""
Hashing in Algorithms — Rabin-Karp, Rolling Hashes
Phase 04 — Algorithms & Complexity Analysis
"""

import random


def polynomial_hash(s: str, base: int = 131, mod: int = 10**9 + 7) -> int:
    h = 0
    for ch in s:
        h = (h * base + ord(ch)) % mod
    return h


def rolling_hash_search(text: str, pattern: str) -> list[int]:
    n, m = len(text), len(pattern)
    if m > n or m == 0:
        return []
    base, mod = 131, 10**9 + 7
    h_pattern = polynomial_hash(pattern, base, mod)
    h_window = polynomial_hash(text[:m], base, mod)
    power = pow(base, m - 1, mod)
    matches = []
    for i in range(n - m + 1):
        if h_window == h_pattern:
            if text[i : i + m] == pattern:
                matches.append(i)
        if i < n - m:
            h_window = (h_window - ord(text[i]) * power) * base + ord(text[i + m])
            h_window %= mod
    return matches


def multi_pattern_search(text: str, patterns: list[str]) -> dict[int, list[int]]:
    base, mod = 131, 10**9 + 7
    by_len: dict[int, set[int]] = {}
    for p in patterns:
        by_len.setdefault(len(p), set()).add(polynomial_hash(p, base, mod))
    results: dict[int, list[int]] = {}
    for length, hash_set in by_len.items():
        if length > len(text):
            continue
        power = pow(base, length - 1, mod)
        h = polynomial_hash(text[:length], base, mod)
        for i in range(len(text) - length + 1):
            if h in hash_set:
                results.setdefault(length, []).append(i)
            if i < len(text) - length:
                h = (h - ord(text[i]) * power) * base + ord(text[i + length])
                h %= mod
    return results


def longest_common_substring(s1: str, s2: str) -> int:
    base, mod = 131, 10**9 + 7

    def has_common(L: int) -> bool:
        if L == 0:
            return True
        hashes: set[int] = set()
        power = pow(base, L - 1, mod)
        h = polynomial_hash(s1[:L], base, mod)
        hashes.add(h)
        for i in range(1, len(s1) - L + 1):
            h = (h - ord(s1[i - 1]) * power) * base + ord(s1[i + L - 1])
            h %= mod
            hashes.add(h)
        h = polynomial_hash(s2[:L], base, mod)
        if h in hashes:
            return True
        for i in range(1, len(s2) - L + 1):
            h = (h - ord(s2[i - 1]) * power) * base + ord(s2[i + L - 1])
            h %= mod
            if h in hashes:
                return True
        return False

    lo, hi, best = 0, min(len(s1), len(s2)), 0
    while lo <= hi:
        mid = (lo + hi) // 2
        if has_common(mid):
            best = mid
            lo = mid + 1
        else:
            hi = mid - 1
    return best


def double_hash(
    s: str,
    p1: int = 131,
    m1: int = 10**9 + 7,
    p2: int = 137,
    m2: int = 10**9 + 9,
) -> tuple[int, int]:
    return (polynomial_hash(s, p1, m1), polynomial_hash(s, p2, m2))


def rabin_karp_double(text: str, pattern: str) -> list[int]:
    n, m = len(text), len(pattern)
    if m > n or m == 0:
        return []
    p1, m1, p2, m2 = 131, 10**9 + 7, 137, 10**9 + 9
    hp = double_hash(pattern, p1, m1, p2, m2)
    hw1 = polynomial_hash(text[:m], p1, m1)
    hw2 = polynomial_hash(text[:m], p2, m2)
    pw1, pw2 = pow(p1, m - 1, m1), pow(p2, m - 1, m2)
    matches = []
    for i in range(n - m + 1):
        if (hw1, hw2) == hp:
            if text[i : i + m] == pattern:
                matches.append(i)
        if i < n - m:
            hw1 = ((hw1 - ord(text[i]) * pw1) * p1 + ord(text[i + m])) % m1
            hw2 = ((hw2 - ord(text[i]) * pw2) * p2 + ord(text[i + m])) % m2
    return matches


def birthday_paradox_simulation(modulus: int, trials: int = 10000) -> dict[int, float]:
    results: dict[int, float] = {}
    for q in [100, 1000, 10000, 100000]:
        collisions = 0
        for _ in range(trials):
            seen: set[int] = set()
            hit = False
            for _ in range(q):
                v = random.randint(0, modulus - 1)
                if v in seen:
                    hit = True
                    break
                seen.add(v)
            if hit:
                collisions += 1
        results[q] = collisions / trials
    return results


def main() -> None:
    text = "abcabcabc"
    pattern = "abc"
    print(f"Rabin-Karp search for '{pattern}' in '{text}':")
    print(f"  Single pattern: {rolling_hash_search(text, pattern)}")
    print(f"  Double hash:    {rabin_karp_double(text, pattern)}")

    patterns = ["abc", "bca", "xyz"]
    print(f"\nMulti-pattern search {patterns}:")
    print(f"  {multi_pattern_search(text, patterns)}")

    s1 = "banana"
    s2 = "canaan"
    print(f"\nLongest common substring of '{s1}' and '{s2}': {longest_common_substring(s1, s2)}")

    # Brute-force verification
    def brute_lcs(a: str, b: str) -> int:
        best = 0
        for i in range(len(a)):
            for j in range(i + 1, len(a) + 1):
                if a[i:j] in b:
                    best = max(best, j - i)
        return best

    assert longest_common_substring(s1, s2) == brute_lcs(s1, s2)

    # Distinct substring count via rolling hash
    def count_distinct_substrings(s: str) -> int:
        base, mod = 131, 10**9 + 7
        seen: set[int] = set()
        for length in range(1, len(s) + 1):
            power = pow(base, length - 1, mod)
            h = polynomial_hash(s[:length], base, mod)
            seen.add(h)
            for i in range(1, len(s) - length + 1):
                h = (h - ord(s[i - 1]) * power) * base + ord(s[i + length - 1])
                h %= mod
                seen.add(h)
        return len(seen)

    print(f"\nDistinct substrings of 'aba': {count_distinct_substrings('aba')} (brute: 5)")

    # Birthday paradox
    print("\nBirthday paradox (M = 10^9+7, 10k trials):")
    bp = birthday_paradox_simulation(10**9 + 7, trials=10000)
    for q, rate in bp.items():
        print(f"  q={q:>6}: collision rate = {rate:.4f}")


if __name__ == "__main__":
    main()

File: phases/04-algorithms-and-complexity/21-hashing-in-algorithms-rabin-karp-rolling-hashes/code/main.rs
//! Hashing in Algorithms — Rabin-Karp, Rolling Hashes
//! Phase 04 — Algorithms & Complexity Analysis

const BASE: u64 = 131;
const MOD: u64 = 1_000_000_007;
const BASE2: u64 = 137;
const MOD2: u64 = 1_000_000_009;

fn polynomial_hash(s: &[u8], base: u64, modulus: u64) -> u64 {
    let mut h: u64 = 0;
    for &ch in s {
        h = (h.wrapping_mul(base).wrapping_add(ch as u64)) % modulus;
    }
    h
}

fn rolling_hash_search(text: &str, pattern: &str) -> Vec<usize> {
    let (t, p) = (text.as_bytes(), pattern.as_bytes());
    let (n, m) = (t.len(), p.len());
    if m > n || m == 0 {
        return vec![];
    }
    let h_pat = polynomial_hash(p, BASE, MOD);
    let mut h_win = polynomial_hash(&t[..m], BASE, MOD);
    let power = mod_pow(BASE, (m - 1) as u32, MOD);
    let mut matches = Vec::new();
    for i in 0..=(n - m) {
        if h_win == h_pat && &t[i..i + m] == p {
            matches.push(i);
        }
        if i < n - m {
            h_win = (MOD + h_win
                - ((t[i] as u64).wrapping_mul(power) % MOD))
                .wrapping_mul(BASE)
                .wrapping_add(t[i + m] as u64)
                % MOD;
        }
    }
    matches
}

fn multi_pattern_search(text: &str, patterns: &[&str]) -> Vec<(usize, usize)> {
    let t = text.as_bytes();
    let mut results = Vec::new();
    let mut by_len: std::collections::HashMap<usize, std::collections::HashSet<u64>> =
        std::collections::HashMap::new();
    for &p in patterns {
        by_len
            .entry(p.len())
            .or_default()
            .insert(polynomial_hash(p.as_bytes(), BASE, MOD));
    }
    for (length, hash_set) in &by_len {
        if *length > t.len() {
            continue;
        }
        let power = mod_pow(BASE, (*length - 1) as u32, MOD);
        let mut h = polynomial_hash(&t[..*length], BASE, MOD);
        for i in 0..=(t.len() - length) {
            if hash_set.contains(&h) {
                results.push((*length, i));
            }
            if i < t.len() - length {
                h = (MOD + h
                    - ((t[i] as u64).wrapping_mul(power) % MOD))
                    .wrapping_mul(BASE)
                    .wrapping_add(t[i + length] as u64)
                    % MOD;
            }
        }
    }
    results
}

fn longest_common_substring(s1: &str, s2: &str) -> usize {
    let (b1, b2) = (s1.as_bytes(), s2.as_bytes());

    fn has_common(b1: &[u8], b2: &[u8], l: usize) -> bool {
        if l == 0 {
            return true;
        }
        let power = mod_pow(BASE, (l - 1) as u32, MOD);
        let mut hashes = std::collections::HashSet::new();
        let mut h = polynomial_hash(&b1[..l], BASE, MOD);
        hashes.insert(h);
        for i in 1..=(b1.len() - l) {
            h = (MOD + h
                - ((b1[i - 1] as u64).wrapping_mul(power) % MOD))
                .wrapping_mul(BASE)
                .wrapping_add(b1[i + l - 1] as u64)
                % MOD;
            hashes.insert(h);
        }
        h = polynomial_hash(&b2[..l], BASE, MOD);
        if hashes.contains(&h) {
            return true;
        }
        for i in 1..=(b2.len() - l) {
            h = (MOD + h
                - ((b2[i - 1] as u64).wrapping_mul(power) % MOD))
                .wrapping_mul(BASE)
                .wrapping_add(b2[i + l - 1] as u64)
                % MOD;
            if hashes.contains(&h) {
                return true;
            }
        }
        false
    }

    let mut lo = 0usize;
    let mut hi = b1.len().min(b2.len());
    let mut best = 0usize;
    while lo <= hi {
        let mid = (lo + hi) / 2;
        if has_common(b1, b2, mid) {
            best = mid;
            lo = mid + 1;
        } else if mid > 0 {
            hi = mid - 1;
        } else {
            break;
        }
    }
    best
}

fn rabin_karp_double(text: &str, pattern: &str) -> Vec<usize> {
    let (t, p) = (text.as_bytes(), pattern.as_bytes());
    let (n, m) = (t.len(), p.len());
    if m > n || m == 0 {
        return vec![];
    }
    let (hp1, hp2) = (
        polynomial_hash(p, BASE, MOD),
        polynomial_hash(p, BASE2, MOD2),
    );
    let mut hw1 = polynomial_hash(&t[..m], BASE, MOD);
    let mut hw2 = polynomial_hash(&t[..m], BASE2, MOD2);
    let pw1 = mod_pow(BASE, (m - 1) as u32, MOD);
    let pw2 = mod_pow(BASE2, (m - 1) as u32, MOD2);
    let mut matches = Vec::new();
    for i in 0..=(n - m) {
        if hw1 == hp1 && hw2 == hp2 && &t[i..i + m] == p {
            matches.push(i);
        }
        if i < n - m {
            hw1 = (MOD + hw1
                - ((t[i] as u64).wrapping_mul(pw1) % MOD))
                .wrapping_mul(BASE)
                .wrapping_add(t[i + m] as u64)
                % MOD;
            hw2 = (MOD2 + hw2
                - ((t[i] as u64).wrapping_mul(pw2) % MOD2))
                .wrapping_mul(BASE2)
                .wrapping_add(t[i + m] as u64)
                % MOD2;
        }
    }
    matches
}

fn mod_pow(mut base: u64, mut exp: u32, modulus: u64) -> u64 {
    let mut result = 1u64;
    base %= modulus;
    while exp > 0 {
        if exp & 1 == 1 {
            result = result.wrapping_mul(base) % modulus;
        }
        exp >>= 1;
        base = base.wrapping_mul(base) % modulus;
    }
    result
}

fn count_distinct_substrings(s: &str) -> usize {
    let b = s.as_bytes();
    let mut seen = std::collections::HashSet::new();
    for length in 1..=b.len() {
        let power = mod_pow(BASE, (length - 1) as u32, MOD);
        let mut h = polynomial_hash(&b[..length], BASE, MOD);
        seen.insert(h);
        for i in 1..=(b.len() - length) {
            h = (MOD + h
                - ((b[i - 1] as u64).wrapping_mul(power) % MOD))
                .wrapping_mul(BASE)
                .wrapping_add(b[i + length - 1] as u64)
                % MOD;
            seen.insert(h);
        }
    }
    seen.len()
}

fn brute_lcs(a: &str, b: &str) -> usize {
    let mut best = 0usize;
    for i in 0..a.len() {
        for j in (i + 1)..=a.len() {
            if b.contains(&a[i..j]) {
                best = best.max(j - i);
            }
        }
    }
    best
}

fn main() {
    let text = "abcabcabc";
    let pattern = "abc";
    println!("Rabin-Karp search for '{pattern}' in '{text}':");
    println!("  Single pattern: {:?}", rolling_hash_search(text, pattern));
    println!("  Double hash:    {:?}", rabin_karp_double(text, pattern));

    let patterns = ["abc", "bca", "xyz"];
    println!("\nMulti-pattern search {patterns:?}:");
    println!("  {:?}", multi_pattern_search(text, &patterns));

    let s1 = "banana";
    let s2 = "canaan";
    let lcs = longest_common_substring(s1, s2);
    println!("\nLongest common substring of '{s1}' and '{s2}': {lcs}");
    assert_eq!(lcs, brute_lcs(s1, s2));

    let distinct = count_distinct_substrings("aba");
    println!("Distinct substrings of 'aba': {distinct} (expected: 5)");
    assert_eq!(distinct, 5);

    println!("\nAll assertions passed.");
}
